new-log-viewer/src/services/decoders/JsonlDecoder.ts

```diff
@@ -0,0 +1,145 @@
+import {
+    Decoder,
+    DecodeResultType,
+    JsonlDecoderOptionsType,
+    LogEventCount,
+} from "../../typings/decoders";
+import {Formatter} from "../../typings/formatters";
+import {JsonObject} from "../../typings/js";
+import {LOG_LEVEL} from "../../typings/logs";
+import LogbackFormatter from "../formatters/LogbackFormatter";
+
+
+/**
+ * A decoder for JSONL (JSON lines) files that contain log events. See `JsonlDecodeOptionsType` for
+ * properties that are specific to log events (compared to generic JSON records).
+ */
+class JsonlDecoder implements Decoder {
+    static #textDecoder = new TextDecoder();
+
+    #logLevelKey: string = "level";
+
+    #logEvents: JsonObject[] = [];
+
+    // @ts-expect-error #fomatter is set in the constructor by `setDecoderOptions()`
+    #formatter: Formatter;
+
+    /**
+     * Parses each line from the given data array as a JSON object and buffers it internally. If a
+     * line cannot be parsed as a JSON object, an error is logged and the line is skipped.
+     *
+     * @param dataArray
+     * @param decoderOptions
+     */
+    constructor (dataArray: Uint8Array, decoderOptions: JsonlDecoderOptionsType) {
+        const isOptionSet = this.setDecoderOptions(decoderOptions);
+        if (false === isOptionSet) {
+            throw new Error(
+                `Initial decoder options are erroneous: ${JSON.stringify(decoderOptions)}`
+            );
+        }
+
+        const text = JsonlDecoder.#textDecoder.decode(dataArray);
+        let beginIdx = 0;
+        while (beginIdx < text.length) {
+            const endIdx = text.indexOf("\n", beginIdx);
+            let line;
+            if (-1 === endIdx) {
+                line = text.substring(beginIdx);
+                beginIdx = text.length;
+            } else {
+                line = text.substring(beginIdx, endIdx);
+                beginIdx = endIdx + 1;
+            }
+
+            try {
+                const logEvent = JSON.parse(line) as JsonObject;
+                this.#logEvents.push(logEvent);
+            } catch (e) {
+                if (0 !== line.length) {
+                    console.error(e, line);
+                }
+            }
+        }
+    }
+
+    getEstimatedNumEvents (): number {
+        return this.#logEvents.length;
+    }
+
+    buildIdx (beginIdx: number, endIdx: number): LogEventCount | null {
+        // This method is a dummy implementation since the actual deserialization is done in the
+        // constructor.
+
+        if (0 > beginIdx || endIdx >= this.#logEvents.length) {
+            return null;
+        }
+
+        return {
+            numValidEvents: endIdx - beginIdx,
+            numInvalidEvents: 0,
```

Comment:
We could technically count the number of invalid events in the constructor and then return something accurate here, right?

Reply:
Implemented below:
1. If a line cannot be parsed as a JSON object, the error is logged, the line is skipped, and a placeholder is inserted into the results. In decode(), such lines are directly inserted without formatting.
2. Add tracking for invalid log events, which now contributes to the total count of invalid events returned by buildIdx().